src/43F0.c

```diff
@@ -74,9 +74,138 @@ HeapNode* _heap_create(s32* addr, u32 size) {
     }
 }
 
-INCLUDE_ASM(s32, "43F0", _heap_malloc);
+#define _HEAP_ALLOC_AND_UPDATE_ID(node) { \
+    u16 HeapEntryID = heap_nextMallocID; \
+    node->allocated = TRUE; \
+    heap_nextMallocID = HeapEntryID + 1; \
+    node->entryID = HeapEntryID; \
+}
+
+void* _heap_malloc(HeapNode* head, u32 size) {
+    HeapNode* nextHeapNode;
+    HeapNode* pPrevHeapNode = NULL;
+    u32 newBlockSize;
+    u32 curBlockLength;
+    HeapNode* curHeapNode;
+    u32 smallestBlockFound;
+
+    // must allocate 16 bytes or more at minimum or fail
+    size = ALIGN16(size);
+    if (!size) {
+        return NULL;
+    }
+
+    smallestBlockFound = 0;
+    nextHeapNode = NULL;
+
+    // find the smallest block we can fit into in the free list
+    for (curHeapNode = head; ; curHeapNode = curHeapNode->next) {
+        if (!curHeapNode->allocated) {
+            curBlockLength = curHeapNode->length;
+            if ((curBlockLength >= size) && (curBlockLength < smallestBlockFound || !smallestBlockFound)) {
+                pPrevHeapNode = curHeapNode;
+                smallestBlockFound = curBlockLength;
+                nextHeapNode = curHeapNode->next;
+            }
+        }
+        if (!curHeapNode->next) {
+            break;
+        }
+    }
+
+
+    // find out the required block size with header
+    newBlockSize = size + sizeof(HeapNode);
+
+    // if we found a block see if we need to split it up
+    if (smallestBlockFound) {
+        if (smallestBlockFound >= newBlockSize) {
+            // update previous to the proper size for the block being returned
+            pPrevHeapNode->next = (HeapNode *)((u8 *)pPrevHeapNode + newBlockSize);
+            pPrevHeapNode->length = size;
+            _HEAP_ALLOC_AND_UPDATE_ID(pPrevHeapNode);
+
+            // setup the new heap block entry
+            curHeapNode = pPrevHeapNode->next;
+            curHeapNode->next = nextHeapNode;
+            curHeapNode->length = smallestBlockFound - newBlockSize;
+            curHeapNode->allocated = FALSE;
+        } else {
+            // take this entry out of the free linked list and mark as allocated
+            pPrevHeapNode->next = nextHeapNode;
+            pPrevHeapNode->length = smallestBlockFound;
+
+            // update the entry id on allocation
+            _HEAP_ALLOC_AND_UPDATE_ID(pPrevHeapNode);
+        }
+        return (u8 *)pPrevHeapNode + sizeof(HeapNode);
```

Comment:
Left-align pointers:
```suggestion
        return (u8*)pPrevHeapNode + sizeof(HeapNode);
```

This might also be better as `return pPrevHeapNode + 1;`?

Reply:
Oops on the space, per the +sizeof, malloc functions (not this one) sometimes return a value in the middle of the struct so it is more common to see some size added instead of doing an array pointer reference. Yes an array reference to [1] would create the same result but I see that as a side effect of it stashing the full struct at the beginning instead of only a partial struct.

Reply:
@ethteck which would you prefer? I'm leaving this and the next comment open for final input.